hi-ml-histopathology/testhisto/testhisto/utils/test_report_utils.py

```diff
@@ -0,0 +1,164 @@
+               return_value=MockHyperDriveRun(child_indices)):
+        crossval_dfs = collect_crossval_outputs(parent_run_id="",
+                                                download_dir=download_dir,
+                                                aml_workspace=None,
+                                                crossval_arg_name=crossval_arg_name,
+                                                output_filename=output_filename)
+
+    assert set(crossval_dfs.keys()) == set(child_indices)
+    assert list(crossval_dfs.keys()) == sorted(crossval_dfs.keys())
+
+    for child_index, child_df in crossval_dfs.items():
+        assert child_df.columns.tolist() == columns
+        assert child_df.loc[0, 'split'] == child_index
+
+
+@pytest.fixture
+def metrics_df() -> pd.DataFrame:
+    return pd.DataFrame({
+        0: {
+            'val/accuracy': [0.3, 0.1, 0.2],
+            'val/auroc': [0.3, 0.1, 0.2],
+            'test/accuracy': 0.3,
+            'test/auroc': 0.3
+        },
+        3: {
+            'val/accuracy': [0.4, 0.5, 0.6],
+            'val/auroc': [0.4, 0.5, 0.6],
+            'test/accuracy': 0.6,
+            'test/auroc': 0.6
+        },
+        1: {
+            'val/accuracy': [0.8, 0.9, 0.7],
+            'val/auroc': [0.8, 0.9, 0.7],
+            'test/accuracy': 0.9,
+            'test/auroc': 0.9
+        }
+    })
+
+
+@pytest.fixture
+def best_epochs(metrics_df: pd.DataFrame) -> Dict[int, int]:
+    return get_best_epochs(metrics_df, 'val/accuracy', maximise=True)
+
+
+@pytest.fixture
+def best_epoch_metrics(metrics_df: pd.DataFrame, best_epochs: Dict[int, int]) -> pd.Series:
+    metrics_list = ['val/accuracy', 'val/auroc']
+    return get_best_epoch_metrics(metrics_df, metrics_list, best_epochs)
+
+
+@pytest.mark.parametrize('maximise', [True, False])
+def test_get_best_epochs(metrics_df: pd.DataFrame, maximise: bool) -> None:
+    best_epochs = get_best_epochs(metrics_df, 'val/accuracy', maximise=maximise)
+    assert list(best_epochs.keys()) == list(metrics_df.columns)
+    assert all(isinstance(epoch, int) for epoch in best_epochs.values())
+
+    expected_best = {0: 0, 1: 1, 3: 2} if maximise else {0: 1, 1: 2, 3: 0}
+    for split in metrics_df.columns:
+        assert best_epochs[split] == expected_best[split]
+
+
+def test_get_best_epoch_metrics(metrics_df: pd.DataFrame, best_epochs: Dict[int, int]) -> None:
+    metrics_list = ['val/accuracy', 'val/auroc']
+    best_metrics_df = get_best_epoch_metrics(metrics_df, metrics_list, best_epochs)
+    assert list(best_metrics_df.index) == metrics_list
+    assert list(best_metrics_df.columns) == list(metrics_df.columns)
+    # Check that all values are now scalars instead of lists:
+    for metric in metrics_list:
+        assert best_metrics_df.loc[metric].map(pd.api.types.is_number).all()
+
+
+def _test_get_crossval_metrics_table(df: pd.DataFrame, metrics_list: List[str]) -> None:
```

Comment:
Can we not use `pytest.mark.parameterize` here to pass in the different metrics lists? (rather than having the 2 extra functions in lines 159-164)?

Reply:
Yes that snippet matches what I had in mind. I personally think it's a lot neater!